File: Sum of Root To Leaf Binary Numbers/Sum of Root To Leaf Binary Numbers.py
# Solution for Sum of Root To Leaf Binary Numbers in PY

class Solution:
    def sumRootToLeaf(self, root) -> int:

        def solve(node, current_value):
            if not node:
                return 0

            # Shift left and add current bit
            current_value = (current_value << 1) + node.val

            # If leaf node
            if not node.left and not node.right:
                return current_value

            return solve(node.left, current_value) + solve(node.right, current_value)

        return solve(root, 0)

File: Sum of Root To Leaf Binary Numbers/Sum of Root To Leaf Binary Numbers.java
// Solution for Sum of Root To Leaf Binary Numbers in JAVA

class Solution {
    private int solve(TreeNode root, int currentValue) {

        if (root == null)
            return 0;

        // Shift left and add current bit
        currentValue = (currentValue << 1) + root.val;

        // If leaf node
        if (root.left == null && root.right == null)
            return currentValue;

        // Recurse left and right
        return solve(root.left, currentValue)
             + solve(root.right, currentValue);
    }

    public int sumRootToLeaf(TreeNode root) {
        return solve(root, 0);
    }
}
